### ISSUE TITLE
create a zoltar option in connect_hub()

### ISSUE THREAD

**Issue by nickreich** [Infectious-Disease-Modeling-Hubs/hubUtils#54]
For starters, I'm hoping that @matthewcornell can help scope out what would be needed here and implement it. There may be some needed changes in the Zoltar system as well to make this work. @annakrystalli do you have any thoughts about what important things to think about here would be, or what hurdles we may face in doing this?

**Comment by annakrystalli** [comment]
I think the biggest hurdle from my still limited understanding of zoltar is that the database structure and storage format (json) in Zoltar is very different from the file structure of file system hubs and what I envisioned for a potential hub database.  My understanding is that it's a result of, among things, solving issues similar to the `type_id` column datatype we have been discussing where a more fine grained approach to preserve datatypes correctly has been developed but which required additional processing to return data from a query.

At this point, I'm not sure whether or how we might be able to mirror a similar connection to Zoltar as we do for file system hubs (local or cloud) though `arrow`, but it is a problem I had earmarked as a very good candidate for me to work on in person with @matthewcornell on my visit.

**Comment by matthewcornell** [comment]
@annakrystalli , without knowing more about this project, I think you've summarized the main issue: different fundamental data storage concepts (and therefore implementations) that can (will?) cause an impedance mismatch. Q: Is there documentation for the API/contract required by the hub interface to the various storage systems (repo, S3, etc)? (Yesterday @nickreich pointed me to [Connect to model output data](https://infectious-disease-modeling-hubs.github.io/hubUtils/reference/connect_hub.html) , but I haven't started digging into anything yet. I figured a conversation with @annakrystalli is a first step.)

**Comment by matthewcornell** [comment]
Again, without having done any research, I'll throw out a naive (and likely untenable) solution: Dump the zoltar database for the project (maybe a subset of locations/targets/etc) into S3 file(s) :-O This is essentially a caching approach, with the attendant issues (when to run, being out of sync, etc.)

**Comment by matthewcornell** [comment]
A longer-term conversation that might be relevant is playing with a Zoltar v.5 data storage change (discussed a while back) where we move the JSON-based data to a separate server (e.g., MongoDB). I wonder if that back end would be more amenable to what we're discussing here...

**Comment by matthewcornell** [comment]
W/ @nickreich and @elray : We might not need to simulate an entire hub structure. Instead maybe we connect and then translate requests to zoltar queries and results to the desired dataframe format.

**Comment by annakrystalli** [comment]
Decision was made **not to try and harmonise zoltar workflow with dplyr filtering**.

Instead zoltar should have connect function in hubUtils but loading data will follow [`load_forecasts_zoltar`](http://reichlab.io/covidHubUtils/reference/load_forecasts_zoltar.html) covidHubUtils paradigm (i.e. through `load_forecasts_zoltar()` function or similar).

### `connect_zoltar()`

Arguments:
- `project_id`: integer
- `host`: defaults to <https://zoltardata.com/>
- `username`: defaults to `Sys.getenv("Z_USERNAME")`
- `password`:  defaults to `Sys.getenv("Z_PASSWORD")`

Output:
  an `zoltr` ZoltarConnection.

```r
connect_zoltar(project_id = 44L)
```

### `load_forecasts_zoltar()`

```r
load_forecasts_zoltar()
```

Keeping arguments consistent with zoltar schema allows for greater flexibility in accessing more that the two hubs that the `covidHubUtils::load_forecasts_zoltar()` used.

The project level API page can be useful for supporting online exploration of potential filtering values as well as for providing functionality (through `zoltr`) for interrogeting project metadata programmatically.


Arguments:
- `conn`: an `zoltr` ZoltarConnection.
- `timezero`: time zero on Zoltar.
- `target`: integer
- `unit`: character

Output:
  a data.frame/tibble of data in long format (consistent with hubverse) with columns:
  - `model_abbr`
  - `team_abbr`
  - `target`: (can horizon be split out when returned @matthewcornell ? )
  - `unit`
  - `season`: if present.
  - `output_type`
  - `type_id` 
  - `value`


### Zoltar to hubverse terminology mapping notes

-  `point` in Zoltar not explicit enough. When passing downloaded data from Zoltar to downstream functions like `ensemble()`, `plot()` etc, its safe-ish enough to translate `point` to `median` and issue warning.
- `bin`: is output type `pmf`
- `cat`: `type_id` for `bin` variables
- `prob`: `value` for `bin` variables
- Season is optional and if present should be passed as a task id.
- Any named distribution data (in `family` and related columns) should be ignored (with warning if some data present).


Useful links:

- https://docs.zoltardata.com/targets/
- https://docs.zoltardata.com/forecastqueryformat/
- https://docs.zoltardata.com/datamodel/

**Comment by matthewcornell** [comment]
**Re: can horizon be split out when returned**: Are you asking if I can remove the `numeric_horizon` from each resulting "full" target name? For example, consider [this target](https://zoltardata.com/api/target/2032/) from the Zoltar [COVID-19 Forecasts project](https://zoltardata.com/project/44):

```json
  {
    "id": 2032,
    "url": "https://zoltardata.com/api/target/2032/",
    "name": "17 day ahead inc hosp",
    "type": "continuous",
    "description": "Zero- to one-hundred-thirty-day ahead forecasts will be defined as the number of incident hospitalizations for the target day.",
    "outcome_variable": "incident hospitalizations",
    "is_step_ahead": true,
    "numeric_horizon": 17,
    "reference_date_type": "DAY",
    "cats": []
  }
```

Here, is the desire to replace "17 day ahead inc hosp" in the Zoltar query output (see attached file) with something like "day ahead inc hosp"? Note that Zoltar has the notion of target "groups", which is obtained for step ahead targets by combining the target's `reference_date_type` and `outcome_variable`. For example, the above target is in the "day ahead incident hospitalizations" group.

Attachment: This is the result of the query `{}` on the Zoltar [Docs Example Project](https://zoltardata.com/project/41). This example contains all types of output available in zoltar.

[mc-0606-docs-query-all-forecasts.csv](https://github.com/Infectious-Disease-Modeling-Hubs/hubUtils/files/11665365/mc-0606-docs-query-all-forecasts.csv)

**Comment by annakrystalli** [comment]
> Here, is the desire to replace "17 day ahead inc hosp" in the Zoltar query output (see attached file) with something like "day ahead inc hosp"? Note that Zoltar has the notion of target "groups", which is obtained for step ahead targets by combining the target's `reference_date_type` and `outcome_variable`. For example, the above target is in the "day ahead incident hospitalizations" group.

Yes, replace `17 day ahead inc hosp` with  `day ahead inc hosp` but add an additional `horizon` field that contains `17`.

**Comment by annakrystalli** [comment]
> Re: **Output**: @annakrystalli do we want `unit` unchanged? Or should it be transformed based on the Hub config task ids or some such?

I'm not 100% sure at the minute. I think what unit can be replaced with might become clearer once I have am example of a hubverse style hub that has been ported to Zoltar. So for now I think `unit` is fine.

**Comment by nickreich** [comment]
@matthewcornell one small modification, is that we on Tuesday decided to just a single `model_id` column instead of two `model_abbr` and `team_abbr` columns, so the zoltar `model` field should be mapped to `model_id` in hubverse.
